### ISSUE TITLE
Android: Fingerprint authentication not implemented yet

### ISSUE THREAD

**Issue by alexanderjarvis** [oblador/react-native-keychain#116]
It's possible to get if the device supports fingerprint authentication with `getSupportedBiometryType()` but the keychain item doesn't use the fingerprint authentication (or provide options for doing so).

Looking at https://github.com/oblador/react-native-keychain/blob/c4b2b4dcfe4e36d2cfc86bf25d3a5af461779464/android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageKeystoreAESCBC.java#L71

The code is commented out. I've tried playing with this feature and uncommenting the code and even passing if the device supports fingerprint auth as an extra argument to encrypt.

However I can't get it working and there's nothing in the logs to suggest what is wrong.

**Comment by oblador** [comment]
@alexanderjarvis It's on the todo, but I'm afraid my Android skills are somewhat lacking and the scope of work is fairly big. Storing it with fingerprint encryption isn't that hard, but unlocking it is a pretty big task as it involves a fair bit of UI and logic on how to handle retries etc which iOS neatly just handles for you. I'd be eternally grateful for help with this and could brief anybody willing to take a stab at it of the complexities involved.

**Comment by alexanderjarvis** [comment]
@oblador thanks for your reply. It's a shame that unlocking it is not so straightforward. If I get some time I may look at doing it and create a PR.

**Comment by pcoltau** [comment]
I would recommend that we don't add any UI to this library, but simply notify (through an event?) the app when authentication is needed and then wait for the app to tell us if that was successful or not and then proceed.

**Comment by FlaviooLima** [comment]
Hope to see this implemented soon :)
With this implemented this package can go to the 🌔  🚀 🚀 🚀 🚀

**Comment by oblador** [comment]
Any takers to help out with this? 🙏

**Comment by jacksontbryan** [comment]
React native touch id has done some work around fingerprint. 

https://github.com/naoufal/react-native-touch-id/tree/master/android

Could this work be leveraged to accelerate this feature?

**Comment by cladjules** [comment]
Let me try to add some native code to Android. See how we could link that up easily and PR.

**Comment by LinusU** [comment]
API Level 28 introduces a complete dialog for this: https://developer.android.com/reference/android/hardware/biometrics/BiometricPrompt

Here is a compat lib: https://github.com/fython/BiometricPromptCompat

I of course found all of this just after implementing it myself 😭 😂 

@oblador what do you think about shipping with this dialog? I could possibly whip up a PR, need this in an app I'm working on...

**Comment by cladjules** [comment]
@oblador @LinusU I have started some work around it, using the current API (FingerPrintManager), which doesn't offer a native UI for it, you have to implement your own.

At least the logic is there, the only problem, for now, is, you have to use to put your fingerprint to Encode as well to Decode. 

https://github.com/cladjules/react-native-keychain

**Comment by LinusU** [comment]
I have a working implementation of UI + logic here: https://github.com/LinusU/react-native-biopass

It's using RSA so that it doesn't require a fingerprint to encode, only to decode.

I'd be interested in PRing that up to this repo if we want to include the UI part here. That way I would get iOS support without implementing it myself ☺️

**Comment by LinusU** [comment]
> I will try and add the BiometricPromptCompat so it uses both Compat and BiometricPrompt code.

That's awesome!

Let me know if you need any help ☺️

**Comment by Jyrno42** [comment]
Since I needed this asap I took the modifications by @cladjules and rebased them to master. Then I did a naive integration with BiometricPromptCompat. Feel free to use it (or not use it).

**Comment by LinusU** [comment]
Great to see things happening here! I would love to see this work upstreamed as soon as possible, and would be happy to help with this.

@oblador what are your thoughts on this? How do we best get this merged back?

**Comment by cladjules** [comment]
I'll try to do a more official implementation of BiometricPrompt, anything I could re-use @Jyrno42 would be great.

Will try to do this weekend.

**Comment by vonovak** [comment]
@LinusU I will be able to do code review for android

**Comment by oblador** [comment]
@LinusU @Jyrno42: I'm sorry that my time is somewhat lacking, so without researching much or even reading this thread attentively I have some thoughts/pointers: 

* Ideally we would be using the fingerprint to encrypt the secret like iOS does. Nothing is currently preventing anyone from doing regular fingerprint auth and then retrieving the secret in your app with mentioned libraries, but it's less secure. Will this `BiometricPrompt` enable us to encrypt/decrypt using the fingerprint?
* Make sure to take localization into account (simplest done by accepting custom strings as input). 
* Super excited to finally get this in! 👏 😄

**Comment by LinusU** [comment]
> Will this `BiometricPrompt` enable us to encrypt/decrypt using the fingerprint?

Yes. My implementation in BioPass only allows reading the value if the fingerprint is successfully scanned, otherwise the Android key system won't hand out the private key.

**Comment by LinusU** [comment]
@Jyrno42 5c00a5d9e08772a2e28bb14255b95adcbd256218 doesn't have anything to do with fingerprint, right? Can you submit that as a separate pull request?

**Comment by cladjules** [comment]
@oblador Ideally we would be using the fingerprint to encrypt the secret like iOS does. Nothing is currently preventing anyone from doing regular fingerprint auth and then retrieving the secret in your app with mentioned libraries, but it's less secure. Will this BiometricPrompt enable us to encrypt/decrypt using the fingerprint?

The implementation of FingerprintManager, does encrypt directly the Keychain entry, so it does hardware encryption.

I suggest I will integrate BiometricPromptCompat which uses both BiometricPrompt and FingerprintManager and integrate to directly encrypt the Cipher.

**Comment by oblador** [comment]
@cladjules Cool, maybe collaborate with @LinusU on his PR?

@LinusU Amazing 💯 😄

**Comment by cladjules** [comment]
@LinusU I think we miss an important point and we would need to pass the Cipher to BiometricPrompt and FingerprintManager using the `FingerprintCrypto.CryptoObject` or `BiometricPrompt.CryptoObject` when calling authenticate (which is always null with the Compat) library.

That would require using the same Cipher to both `encrypt` that we can retrieve in the Fingerprint callback.

I will update the PR, but we definitely need to add some bits to the `Compat` one too.

**Comment by Jyrno42** [comment]
> @Jyrno42 [5c00a5d](https://github.com/oblador/react-native-keychain/commit/5c00a5d9e08772a2e28bb14255b95adcbd256218) doesn't have anything to do with fingerprint, right? Can you submit that as a separate pull request?

Will do

**Comment by jordanfloyd** [comment]
Also interested in this. It would be nice to use a single library, but I can use another Fingerprint auth library if needed.

**Comment by vonovak** [comment]
please track progress in https://github.com/oblador/react-native-keychain/pull/148

**Comment by OleksandrKucherenko** [comment]
https://github.com/oblador/react-native-keychain/pull/260 - PR with implemented and tested biometric. Enjoy
